src/Development/IDE/Core/Service.hs

```diff
@@ -85,21 +86,21 @@ shutdown = shakeShut
 runAction :: IdeState -> Action a -> IO a
 runAction ide action = do
     bar <- newBarrier
-    res <- shakeRun ide [do v <- action; liftIO $ signalBarrier bar v; return v]
+    res <- shakeEnqueue ide (do v <- action; liftIO $ signalBarrier bar v; return v)
     -- shakeRun might throw an exception (either through action or a default rule),
     -- in which case action may not complete successfully, and signalBarrier might not be called.
     -- Therefore we wait for either res (which propagates the exception) or the barrier.
     -- Importantly, if the barrier does finish, cancelling res only kills waiting for the result,
     -- it doesn't kill the actual work
-    fmap fromEither $ race (head <$> res) $ waitBarrier bar
+    fmap fromEither $ race res $ waitBarrier bar
```

Comment:
I wonder if we can simplify `runAction` further. Does anyone remember what's the purpose of the Barrier here? Perhaps  it's not  needed anymore with `shakeEnqueue`.

Reply:
shakeRun was asyncronous so the barrier was required to turn it back into a synchronous operation. It seems runInShakeSession already has an IO that can be joined to get the value back, so I'd do that.

Reply:
it's done